### PR TITLE
bpo-38439: Add 256px IDLE icon

### PR DESCRIPTION
**Pull request by hroncok**
Icon author: Andrew Clover, [bpo-1490384](https://bugs.python.org/issue1490384)

Downloaded from https://www.doxdesk.com/software/py/pyicons.html

I would add the vector version as well, however:

 - found now way to open the Xara file
 - the SVG file is missing important features (e.g. shadow is broken)

https://bugs.python.org/issue38439

### CODE REVIEW

**Comment by vstinner** [comment]
cc @terryjreedy @csabella @serhiy-storchaka 

The icon looks great! It looks the same than Lib/idlelib/Icons/idle_48.png but with a much better resolution. I cannot see 256x256 pixels icon in https://bugs.python.org/issue1490384 : where does the file come from? I would like to get some more context to check the license. On the issue, I see that Andrew Clover signed the Contributor Agreement. Did Andrew draw them?

---

The commit 4ade2d25fc585876540ac9d591a44469fd853462 of https://bugs.python.org/issue20406 added the following icons (GIF and PNG formats):

```
 Lib/idlelib/Icons/idle.ico    | Bin 0 -> 19790 bytes
 Lib/idlelib/Icons/idle_16.gif | Bin 0 -> 1034 bytes
 Lib/idlelib/Icons/idle_16.png | Bin 0 -> 1264 bytes
 Lib/idlelib/Icons/idle_32.gif | Bin 0 -> 1435 bytes
 Lib/idlelib/Icons/idle_32.png | Bin 0 -> 2542 bytes
 Lib/idlelib/Icons/idle_48.gif | Bin 0 -> 1388 bytes
 Lib/idlelib/Icons/idle_48.png | Bin 0 -> 4710 bytes
```

But I didn't understand where these files come from.

File: Misc/NEWS.d/next/IDLE/2019-12-05-14-20-53.bpo-38439.j_L2PI.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Add a 256×256 pixel IDLE icon to support more modern environments.
```
Comment (vstinner) [inline Misc/NEWS.d/next/IDLE/2019-12-05-14-20-53.bpo-38439.j_L2PI.rst]: Mention maybe that Andrew Clover drew them.

Comment (hroncok) [reply Misc/NEWS.d/next/IDLE/2019-12-05-14-20-53.bpo-38439.j_L2PI.rst]: Sure.

**Comment by hroncok** [comment]
> But I didn't understand where these files come from.

I've tried to address this in the commit message.

**Comment by vstinner** [comment]
Oh wait, I missed https://bugs.python.org/issue38439 I confirm that the icon is drew by drew by Andrew Clover: https://bugs.python.org/issue38439#msg357852

For an unknown reason, the bot didn't add a link to https://bugs.python.org/issue38439 in the PR description.

**Comment by hroncok** [comment]
As a side note, I haven't added a gif, because I don't see a point in a 256px gif icon.

**Comment by vstinner** [comment]
I installed Xara Photo & Graphic Designer trial on Windows. I converted the Xara file to SVG at 300 dpi, but the compressed SVG is still 1.4 MiB which is quite big. I converted to 96 dpi compressed SVG: 796 KiB. Still quite big.

* http://haypo.alwaysdata.net/tmp/pyicons-0.4-300dpi.svgz
* http://haypo.alwaysdata.net/tmp/pyicons-0.4-96dpi.svgz

I edited the SVG to only keep the 4 IDLE icons: 152 KiB. That's more reasonable, no?

* http://haypo.alwaysdata.net/tmp/idle.svgz

If I only keep the IDLE icon "at 256x256 pixels" resolution (the one used to generated the PNG of this PR), it's 136 KiB:

* http://haypo.alwaysdata.net/tmp/idle-256px.svgz

I don't have a strong opinion about only adding the PNG, or also add the SVG.

**Comment by hroncok** [comment]
I would also add the svg, possibly with all the variants. At least include t in the sources (not install it). So we don't have to search the Internets when we add a higher resolution icon in next decade.

However, no idea where to put it in the source tree.

**Comment by hroncok** [comment]
And here is the scaleable icon I'd like to actually include int he installation (uncompressed).
[idle-256px.svg.tar.gz](https://github.com/python/cpython/files/3927351/idle-256px.svg.tar.gz)

**Comment by vstinner** [comment]
> And here is the scaleable icon I'd like to actually include in the installation (uncompressed). idle-256px.svg.tar.gz

Uncompressed: 136 KiB. That sounds reasonable. I like the idea of installing it. But I'm not sure of the location. Do you expect to get it in /usr/share? Currently, Python only install icons in /usr/lib, no? I'm not sure how to modify the Makefile. Maybe start by adding the SVG in Lib/idlelib/Icons/ but don't install it with "make install". It's nice to have the source of PNG icons in the same directory.

Linux vendors like Fedora could use this SVG icon if they want.

Or maybe propose a following PR to install it, if you are motivated to hack the Makefile ;-)

**Comment by hroncok** [comment]
In Fedora, we currently do:

```
install -D -m 0644 Lib/idlelib/Icons/idle_16.png %{buildroot}%{_datadir}/icons/hicolor/16x16/apps/idle3.png
install -D -m 0644 Lib/idlelib/Icons/idle_32.png %{buildroot}%{_datadir}/icons/hicolor/32x32/apps/idle3.png
install -D -m 0644 Lib/idlelib/Icons/idle_48.png %{buildroot}%{_datadir}/icons/hicolor/48x48/apps/idle3.png
```

Where `%{buildroot}%{_datadir}` is `/usr/share`.

**Comment by hroncok** [comment]
How can move this forward? Is the higher resolution icon desired? Should I add the SVG "source" as well as the minimized scalable SVG? Where?

We can carry those downstream, but I'd rather not.

**Comment by terryjreedy** [comment]
I am on holiday break with my family, but I will try to get to this, at least from the IDLE side, soon.

**Comment by hroncok** [comment]
Enjoy the break and don't worry. It's enough for me to know that you are aware of this. Thanks for the heads up.

**Review by vstinner — APPROVED** [review]
LGTM. I'm fine with not adding the source SVG into Python, only PNG.

Note: the idle.ico has been modified to add a 8th icon into the file: the 256x256 pixels PNG flavor.

For the one who merges the change: please just ensure that the commit message is complete, see https://github.com/python/cpython/pull/17473#issue-349438316

**Comment by zooba** [comment]
Could someone with the SVG also create 44x44 and 150x150 sized versions? (Like the `pythonx44.png` and `pythonx150.png` files in `PC/icons`.)

These are the sizes used for the tiles created by the Microsoft Store installer, and referenced from `PC\layout\support\appxmanifest.py` (search for `pythonwx` to find the current IDLE and pythonw references).

**Comment by hroncok** [comment]
I cannot do it right now, but the SVG is attached here: https://github.com/python/cpython/files/3927351/idle-256px.svg.tar.gz

**Comment by hroncok** [comment]
and this one has scalable versions of the smaller icons, more likely to be useful for the 44x44 version:  http://haypo.alwaysdata.net/tmp/idle.svgz

**Comment by hroncok** [comment]
@terryjreedy Should we still wait for your review?

**Comment by hroncok** [comment]
44x44:

idle-44px

150x150:

idle-150px

**Review by terryjreedy — CHANGES_REQUESTED** [review]
As is, the patch seems incomplete and it is not clear what effects it will have.  (Revised from first post.)  The relevant code in idlelib.pyshell is
```
    # set application icon
    icondir = os.path.join(os.path.dirname(__file__), 'Icons')
    if system() == 'Windows':
        iconfile = os.path.join(icondir, 'idle.ico')
        root.wm_iconbitmap(default=iconfile)
    elif not macosx.isAquaTk():
        ext = '.png' if TkVersion >= 8.6 else '.gif'
        iconfiles = [os.path.join(icondir, 'idle_%d%s' % (size, ext))
                     for size in (16, 32, 48)]
        icons = [PhotoImage(master=root, file=iconfile)
                 for iconfile in iconfiles]
        root.wm_iconphoto(True, *icons)
```
The patch has an expanded multi-icon idle.ico for Windows.  Since it will only be used on Vista+, It should have been made from the 4 .pngs.  Was it?

An expanded multi-icon idle.icns is needed for macOS Aqua, made from the 4 .pngs.  The mac installer must separately grab idle.icns and give it to the OS for whatever use it makes of it.

For *nix, either idle_256.gif must be added, and 256 added to the list in the code above, or the code needs revision to tolerate its absence. 

Except on macOS Aqua, I am not sure if the new size will ever by used, as the code above only affects the little icon in upper-left corners of windows.   (After I posted this, I notice that Fedora also grabs the icons from IDLE for use elsewhere.)

I don't know what the Content include lines do.  idle.icns is installed even though not 'Content include'd.  The .gif and .png files are not directly used, at least not by IDLE.  But worrying about this is another issue.

It seems to me that since installing tkinter/IDLE is generally optional, icons needs by other than IDLE should be stored somewhere else always installed, and 'imported' as needed by IDLE.

**Comment by hroncok** [comment]
> The patch has an expanded multi-icon idle.ico for Windows. Since it will only be used on Vista+, It should have been made from the 4 .pngs. Was it?

No, I've just added the new size. Will revisit this.

> An expanded multi-icon idle.icns is needed for macOS Aqua, made from the 4 .pngs. The mac installer must separately grab idle.icns and give it to the OS for whatever use it makes of it.

I have literal no idea how to do this.

> For *nix, either idle_256.gif must be added, and 256 added to the list in the code above, or the code needs revision to tolerate its absence.

Will try see how the 256px gif looks lik and if it is horrible, will adapt the code.

> It seems to me that since installing tkinter/IDLE is generally optional, icons needs by other than IDLE should be stored somewhere else always installed, and 'imported' as needed by IDLE.

I don't know how can I address this. Would you be able to guide me or should I just create the pngs and let you do the magic?

**Comment by hroncok** [comment]
> The patch has an expanded multi-icon idle.ico for Windows. Since it will only be used on Vista+, It should have been made from the 4 .pngs. Was it?

It was now.

> For *nix, either idle_256.gif must be added, and 256 added to the list in the code above, or the code needs revision to tolerate its absence.

Code adapted, gif was ugly.

**Comment by terryjreedy** [comment]
@ned-deily  We have a new idlelib/icons/idle_256.png to add to the existing masOS idle.icns (or create a new from from all the .pngs).  Can you do that, or suggest the best way to do so?  I presume that idle.icns is being used by the macOS installer.

**Comment by hroncok** [comment]
@terryjreedy I've recreated the .ico in GIMP, from the PNGs. It's just the mac icns I don't know ho to edit.

**Comment by terryjreedy** [comment]
I would like to add icons/README.txt with the information in my response + how you made the .ico + however we make the new .icns file + ???.

**Comment by hroncok** [comment]
I have made the requested changes; please review again.

**Comment by hroncok** [comment]
Hello. I'd like to get this merged before the fist 3.9 beta. Is that realistic, or shall we ship the hi-res icon in Fedora differently?

**Comment by terryjreedy** [comment]
This was near the top of my PR review queue when my development system went 'psst' last Saturday.  Non-boot is still not completely diagnosed.

The code change only affect non-Aqua *nix systems.  It is straightforward and looks correct.  I presume you have verified that it works on Fedora.

Windows: IDLE is only affected by the new .ico.  I would like it tested by running IDLE and seeing if the small icon still appears on IDLE windows.  The PCbuild change is obvious, but does not, I believe, affect IDLE.

@ned-deily @taleinat Can one of you verify that the new .icns works properly?  Is it used for a repository build or only after installation?

.0a6 is scheduled for 4-22 and .0b1 on 5-18.  I would be willing to merge a  partial PR with just the code and project changes and new .png before either.  If we go this route, could you upload the .png to the issue so I can look at it?

IDLE is not normally bound by the usual type and backport rules (see PEPE 434), but that would not necessarily apply to PCbuild, so I also would like the minimum in before .0b1.

**Comment by ned-deily** [comment]
> @ned-deily @taleinat Can one of you verify that the new .icns works properly? Is it used for a repository build or only after installation?

As usual, it's not that simple (sigh!).  After diving back in and refreshing my memory, as far as I can tell, `Lib/idlelib/Icons/idle.icns` is not used anywhere and hasn't been since at least 2006, probably earlier, all long before my time.  In ff8c0221237e345455e6e55b8e4f88118a84e455, a separate set of macOS-specific icons for various objects was added, including ones for IDLE app and .py files.  These are currently in `Mac/IDLE/Icons`.   This is the IDLE app icon:

IDLE

The current Mac-specific icons have a maximum resolution of 128x128 (72 dpi) which still look fairly OK on current hi-res displays but, at some point, it would be good to update them all to provide all [the current Apple recommended resolutions](https://developer.apple.com/design/human-interface-guidelines/macos/icons-and-images/app-icon/) (note the Folder icon was updated separately a couple of years ago in GH-1780).  But that's all definitely out of scope for this PR.  I'll get around to opening an issue for it sometime if no one else does it first.

So the best thing would be for this PR to simply remove the obsolete .icns file rather than try to update it.  Thanks for the effort, @hroncok!

**Comment by hroncok** [comment]
> I would be willing to merge a partial PR with just the code and project changes and new .png before either. If we go this route, could you upload the .png to the issue so I can look at it?

From the list of current changes, what do you mean by "just the code and project changes"? Skipping the ico and icns change?

The PNG:

idle_256

**Comment by terryjreedy** [comment]
@taleinat Ned disposed of the .icns issue.  Do you currently have a respository on Windows?  If so, can you verify that IDLE runs with the current patch and new .ico file?

Ned suggested removing rather than replacing the unused .icns file.  README.txt needs a corresponding change.  That just leaves the .ico file to be held back until verified by someone.

**Comment by hroncok** [comment]
I've deleted the icns file and adapted the news entry and the README.

**Comment by terryjreedy** [comment]
Miro, delete the new .ico file, leaving the old one as is, and I will merge this.  
Then make a 2nd PR that just replaces the .ico, and I will deal with it separately at a later date.

**Comment by taleinat** [comment]
I have a Windows setup and can test this later today.

**Comment by terryjreedy** [comment]
Mike, I received an email about .ico removal, but github says

> We went looking everywhere, but couldn’t find those commits.
> Sometimes commits can disappear after a force-push. Head back to the latest changes here.

When changing a PR, update merges are usually better force pushes.  In any case, Tal responded that he will test the .ico.  If we happen to miss .0a6, we should get this in for .0b1 immediately after.

**Comment by hroncok** [comment]
github has problems today. my commits are in the branch but not visible here.

https://github.com/hroncok/cpython/commits/idle256

I can drop the last one if needed.

**Comment by hroncok** [comment]
The commit is finally here.

**Review by terryjreedy — APPROVED** [review]

**Comment by terryjreedy** [comment]
Please make a 2nd PR, same issue, for the .ico.  Even if the 256 icon is not used, the png versions of the others may improve on the .gif versions.

**Comment by hroncok** [comment]
https://github.com/python/cpython/pull/19648

**Comment by hroncok** [comment]
@stratakis is that ours? Hardly related here.

**Comment by stratakis** [comment]
@hroncok indeed it is, but the failure seems unrelated.

**Comment by vstinner** [comment]
> @stratakis is that ours? Hardly related here.

Sadly, there are many known buildbot failures. Search for bugs that I reopened which contain "test" in the title to have an idea :-)

This one is likely this bug: https://bugs.python.org/issue39995 Reported one month ago, nobody investigated it yet.

**Comment by taleinat** [comment]
> @taleinat Ned disposed of the .icns issue. Do you currently have a respository on Windows? If so, can you verify that IDLE runs with the current patch and new .ico file?

Tested now, works great on my Windows 10 machine.

**Pull request merged**